### PR TITLE
Add support for custom operators

### PR DESCRIPTION
**Pull request by rakbladsvalsen**
## PR Info

See SeaORM discord 

## New Features

- Add support for arbitrary custom operators
- Add `Regex` op for Postgres
- Add `RegexCaseInsensitive` op for Postgres

## Notes

This PR is still a little bit rough around the edges, but hey, I'm open to comments. :)

### CODE REVIEW

**Comment by rakbladsvalsen** [comment]
Some questions: 

- Should I add a new type for custom operators? Currently I just used `Alias` (since it implements `Into Iden`), but using "alias" for a custom operator feels weird. Maybe a new `CustOp` type would make sense?
- I nuked the `Copy` trait from `BinOper`, but this has the downside of adding a clone call somewhere. This is my first time contributing to sea-query, and I'm rather ashamed to ask this: Is there an easy way to make `Ident` `Copy`? I tried adding the bound but I epicly failed at it.

File: src/types.rs [hunk 1]
```diff
@@ -72,6 +72,8 @@ impl Clone for SeaRc<dyn Iden> {
     }
 }
 
+impl Eq for SeaRc<dyn Iden> {}
```
Comment (billy1624) [inline src/types.rs:75]: Hmmmm... `Eq` for `dyn Iden`

Consider this. Where both `dyn Iden` yield the same identifier `id` but it's not the same Rust structure / enum.

```rs
#[derive(Iden)]
struct Id;

// Both yield the string "id"
assert_eq!(Id.to_string(), "id");
assert_eq!(Alias::new("id").to_string(), "id");

// How about this??
assert_eq!(Id, Alias::new("id"));
```

**Review by billy1624 — COMMENTED** [review]
Hey @rakbladsvalsen, thanks for the contribution!! I got a few comments :)

File: src/types.rs (hunk 1 above)
Comment (rakbladsvalsen) [reply src/types.rs:75]: Isn't this already covered by the `PartialEq` implementation that already exists for `DynIden`?  The current implementation checks two things: whether the .to_string() contents are equal and whether the vtable pointer is the same for the two objects, though this only works for dyn objects.

Anyway, if we wanted to provide the ability to do something like the last line in your example, it'd be necessary to implement something like `fn get_iden_id()` to allow users direct comparison between Iden objects, i.e. `Id.get_iden_id() == Alias::new("id").get_iden_id()` (which is probably already doable using `to_string()`). 

Though I must say, this feels rather unnatural. I don't think comparison between two different struct types is something a lot of users are doing. There are no feature requests for this either, so I think we're good in that regard.

Comment (billy1624) [reply src/types.rs:75]: @tyt2y3 thoughts?

File: src/types.rs [hunk 2]
```diff
@@ -158,7 +160,7 @@ pub enum UnOper {
 }
 
 /// Binary operator
-#[derive(Debug, Clone, Copy, PartialEq, Eq)]
+#[derive(Debug, Clone, PartialEq, Eq)]
```
Comment (tyt2y3) [inline src/types.rs:163]: We should not remove the Copy attribute.

File: src/types.rs [hunk 3]
```diff
@@ -185,6 +187,7 @@ pub enum BinOper {
     RShift,
     As,
     Escape,
+    Custom(DynIden),
```
Comment (tyt2y3) [inline src/types.rs:190]: My question is why we would like to reuse DynIden here?
It was originally intended for 'identifiers'.
May be we can have a new type CustomOper, which is simply a wrapped `&'static str`

Comment (billy1624) [reply src/types.rs:190]: Right, I was not aware we need to remove the `Copy` derive. So, yes, `&'static str` is a better alternative.

@rakbladsvalsen what do you think?

Comment (rakbladsvalsen) [reply src/types.rs]: Thanks for the feedback guys! This PR _should_ be good to go.

**Review by billy1624 — APPROVED** [review]
Thanks!! @rakbladsvalsen

**Review by ikrivosheev — APPROVED** [review]
@rakbladsvalsen thank you! LGTM!

**Review by tyt2y3 — APPROVED** [review]
Awesome, looks neat!

**Pull request merged**